docs/docs/writing-plugins/the-rules-api/concepts.mdx

```diff
@@ -23,176 +23,251 @@ async def int_to_str(i: int) -> str:
+- Computed from other params by (transitively) applying suitable rules. You can think of this as a form of dependency injection via type: Pants knows the type of the implicit parameter, and can traverse a path through rule execution to go from an initial set of values, known from context, to the needed value.
+
+Since explicit params must be provided positionally, they must be the first arguments to the rule. This means that when you write a rules, you should put the parameters expected to be passed explicitly before the parameters expected to be provided implicitly.
 
-In this example, we could not have requested the type `ProcessResult` as a parameter to our rule because we needed to dynamically create a `Process` object.
+### Extra context for implicit parameters
 
-Thanks to `await Get`, we can write a recursive rule to compute a [Fibonacci number](https://en.wikipedia.org/wiki/Fibonacci_number):
+As mentioned above, Pants can compute values for implicit parameters by transitively applying rules. In many cases the initial parameters for *those* rules are known from external context. But in some cases we need to provide extra context from the calling rule. To do so, we pass the contextual parameters as arguments to `**implicitly()`:
 
 ```python
-@dataclass(frozen=True)
-class Fibonacci:
-    val: int
+from pants.engine.process import fallible_to_exec_result_or_raise
+from pants.engine.rules import implicitly, rule
 
 @rule
-async def compute_fibonacci(n: int) -> Fibonacci:
-    if n < 2:
-        return Fibonacci(n)
-    x = await Get(Fibonacci, int, n - 2)
-    y = await Get(Fibonacci, int, n - 1)
-    return Fibonacci(x.val + y.val)
+async def my_rule() -> MyResult:
+    process_result = await fallible_to_exec_result_or_raise(
+        **implicitly(
+            Process(
+                ["/bin/echo", str(target.address)],
+                description=f"Echo {target.address}",
+            )
+        )
+    )
+    ...
 ```
 
-Another rule could then "call" our Fibonacci rule by using its own `Get`:
+In this example the `fallible_to_exec_result_or_raise()` rule takes a `FallibleProcessResult` and returns a `ProcessResult` by first checking the `FallibleProcessResult` for success and raising an exception if it failed. We saw this earlier, in the simplified shellcheck example.
 
-```python
-@rule
-async def call_fibonacci(...) -> Foo:
-    fib = await Get(Fibonnaci, int, 4)
-    ...
-```
+But instead of explicitly passing a `FallibleProcessResult` as we did earlier, we now pass a `Process` as implicit context. The Pants engine then looks at all the rules it knows about to figure out how to compute a `FallibleProcessResult` from a `Process`. The `execute_process()` we encountered earlier fits the bill, and so the engine calls it on our `Process` and passes its return value into `fallible_to_exec_result_or_raise()`. Whereas earlier we called both rules explicitly, here we get the exact same behavior with just one call.
```

Comment:
```suggestion
But instead of explicitly passing a `FallibleProcessResult` as we did earlier, we now pass a `Process` as implicit context. The Pants engine then looks at all the rules it knows about to figure out how to compute a `FallibleProcessResult` from a `Process`. The `execute_process()` we encountered earlier fits the bill, and so the engine calls it on our `Process` and passes its return value into `execute_process_or_raise()`. Whereas earlier we called both rules explicitly, here we get the exact same behavior with just one call.
```